### PR TITLE
frontend: useKubeObject: probe endpoints with GET-by-name instead of List

### LINKED ISSUE #4674: [Bug] useGet fails when user has 'get' permission but lacks 'list' permission
## Describe the bug

Headlamp's `useGet` (via `useKubeObject`) can fail to retrieve resource details when a user has `get` permission but does not have `list` permission for that resource.

The current implementation of `getWorkingEndpoint` probes the collection endpoint (for example, `/apis/apps/v1/namespaces/default/deployments`) using a GET request without a resource name. In Kubernetes RBAC, this is authorized as the `list` verb. As a result, users with only `get` permission receive `403 Forbidden` during this pre-check, even though they are authorized to `get` the specific resource by name.

**Expected outcome:**  
Users with `get` permission should be able to open resource details even without `list` permission, similar to `kubectl get <resource> <name>` behavior.

## To Reproduce

Steps to reproduce the bug:

1. Create a Role with get permission for a resource (e.g., services) but without list.
2. Bind the Role to a user and log in to Headlamp.
3. Open the details page of a specific resource.
4. The page fails to load due to a 403 error from the collection URL probe.

## Environment (please provide info about your environment):

- Installation type: Helm (In-Cluster)
- Headlamp Version: 0.39.0
- Other: Kubernetes RBAC enabled.

## Are you able to fix this issue?

Yes (I will propose a PR)

## Additional Context

- Root Cause: useKubeObject() -> useEndpoints() -> getWorkingEndpoint() performs a clusterFetch against a collection URL, which K8s maps to the list verb.
- Affected Resources: Primarily resources with multiple API version candidates (e.g., Ingress, Gateway, or CRDs) where useEndpoints performs preflight probes.
- Comparison: Tools like kubectl and ArgoCD successfully handle this by targeting the name-specific GET path directly.
- Scenario: This is critical in multi-tenant environments where list access to certain resources/CRDs is restricted for security reasons.

### PR DESCRIPTION
**Pull request by Kaniikura**
## Summary

This PR fixes `useKubeObject` failing when a user has `get` permission but lacks `list` permission, by switching endpoint probing from collection-URL LIST to name-based GET for single-resource views.

## Related Issue

Fixes #4674

## Changes

- Updated `useKubeObject` to pass `name` and `namespace` to `useEndpoints`, so endpoint discovery uses a name-specific GET request (`/deployments/demo`) instead of a collection GET (`/deployments`) that Kubernetes authorizes as the `list` verb
- Added `getFirstWorkingGetEndpoint` — parallel GET-by-name probing strategy for single-resource views
- Renamed the existing probing logic to `getFirstWorkingListEndpoint` for clarity
- Added hooks.test.tsx with 7 tests covering both probing strategies and edge cases

## Steps to Test

1. Create a Kubernetes Role granting only `get` (not `list`) permission on a resource (e.g., Deployments) and bind it to a user
2. Log in to Headlamp as that user
3. Navigate directly to the detail page of a specific resource (e.g., `/c/my-cluster/apps/v1/Deployment/default/my-deployment`)
4. Verify the page loads successfully without a 403 error
5. Verify that list views (e.g., Deployments list) still work correctly for users who have `list` permission

## Screenshots (if applicable)

N/A — no UI changes; the fix is in the endpoint probing logic.

## Notes for the Reviewer

- The existing `listProbe` mode (used by `useKubeObjectList`) is **unchanged in behavior** — the new `getProbe` mode is only used by `useKubeObject` for single-resource detail views

### CODE REVIEW

File: frontend/src/lib/k8s/api/v2/hooks.ts [hunk 1]
```diff
@@ -216,44 +218,180 @@ export function useKubeObject<K extends KubeObject>({
...
  *
- * @params endpoints - List of possible endpoints
+ * @param endpoints - List of possible endpoints
  */
 export const useEndpoints = (
   endpoints: KubeObjectEndpoint[],
   cluster: string,
-  namespace?: string
+  namespace?: string,
+  { mode = 'listProbe', name }: UseEndpointsOptions = {}
```
Comment (sniok) [inline frontend/src/lib/k8s/api/v2/hooks.ts]: there's no need for a `mode` paramater. we can just have an optional `name` parameter. if it's present then do individual 'get' request, if it's not present then 'list' request.

File: frontend/src/lib/k8s/api/v2/hooks.ts [hunk 2]
```diff
@@ -216,44 +218,180 @@ export function useKubeObject<K extends KubeObject>({
...
+  clusterName: string,
   namespace?: string
 ) => {
   const promises = endpoints.map(endpoint => {
     return clusterFetch(KubeObjectEndpoint.toUrl(endpoint, namespace), {
       method: 'GET',
-      cluster: cluster ?? getCluster() ?? '',
+      cluster: clusterName,
+    }).then(response => {
+      if (!response.ok) {
```
Comment (sniok) [inline frontend/src/lib/k8s/api/v2/hooks.ts]: there's no need for this check, if response.ok is false it will already throw an ApiError. check out `fetch.ts:51`

File: frontend/src/lib/k8s/api/v2/hooks.ts [hunk 3]
```diff
@@ -216,44 +218,180 @@ export function useKubeObject<K extends KubeObject>({
...
+const toProbeApiError = (error: unknown): ApiError =>
+  error instanceof ApiError
+    ? error
+    : new ApiError(error instanceof Error ? error.message : 'Unreachable');
+
+/**
+ * Pick one actionable user-facing error from probe results.
+ * Prioritize RBAC denial (403) over endpoint mismatches (404/405).
+ */
+const throwMostRelevantError = (errors: unknown[]): never => {
```
Comment (sniok) [inline frontend/src/lib/k8s/api/v2/hooks.ts]: I don't think this is necessary. Case when all endpoints don't work is quite rare and we shouldn't focus on it too much, most of the time one of the endpoints will work.

Comment (Kaniikura) [reply frontend/src/lib/k8s/api/v2/hooks.ts]: Makes sense, removed the prioritization logic.

File: frontend/src/lib/k8s/api/v2/hooks.ts (hunk 1 above)
Comment (Kaniikura) [reply frontend/src/lib/k8s/api/v2/hooks.ts]: Agreed, done.

File: frontend/src/lib/k8s/api/v2/hooks.ts [hunk 4]
```diff
@@ -216,44 +218,180 @@ export function useKubeObject<K extends KubeObject>({
...
+    if (probeError.status === 403) {
+      forbiddenError ??= probeError;
+    } else if (probeError.status === 404 || probeError.status === 405) {
+      mismatchError ??= probeError;
+    } else {
+      otherError ??= probeError;
+    }
+  }
+
+  throw forbiddenError ?? otherError ?? mismatchError ?? new ApiError('Unreachable');
```
[the remark this answers is not in the document]
Comment (Kaniikura) [reply frontend/src/lib/k8s/api/v2/hooks.ts]: this function has been deleted

File: frontend/src/lib/k8s/api/v2/hooks.ts [hunk 5]
```diff
@@ -216,44 +218,180 @@ export function useKubeObject<K extends KubeObject>({
...
+  endpoints: KubeObjectEndpoint[],
+  clusterName: string,
+  name: string,
+  namespace?: string
+) => {
+  const promises = endpoints.map(endpoint => {
+    return clusterFetch(makeUrl([KubeObjectEndpoint.toUrl(endpoint, namespace), name]), {
+      method: 'GET',
+      cluster: clusterName,
     }).then(() => endpoint);
```
[the remark this answers is not in the document]
Comment (Kaniikura) [reply frontend/src/lib/k8s/api/v2/hooks.ts:245]: This is already handled: `clusterFetch` calls `backendFetch`, which throws `ApiError` for any non-ok response before returning. A fulfilled `clusterFetch` promise guarantees `response.ok === true`, so an additional check is unnecessary. Note that `getFirstWorkingListEndpoint` doesn't check `response.ok` either.

[the remark this answers is not in the document]
Comment (illume) [reply frontend/src/lib/k8s/api/v2/hooks.ts:245]: Thanks for the explanation. Marking this as resolved.

File: frontend/src/lib/k8s/api/v2/hooks.test.tsx [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    });
+
+    const { result } = renderHook(() => useEndpoints(endpoints, 'cluster-a', 'default'), {
+      wrapper: createWrapper(),
+    });
+
+    await waitFor(() => {
+      expect(result.current.endpoint).toEqual(endpoints[1]);
+    });
+  });
```
[the remark this answers is not in the document]
Comment (Kaniikura) [reply frontend/src/lib/k8s/api/v2/hooks.test.tsx:276]: I will include a test case for cluster-scoped resources.

[the remark this answers is not in the document]
Comment (Kaniikura) [reply frontend/src/lib/k8s/api/v2/hooks.test.tsx:276]: done.

**Comment by Kaniikura** [comment]
Thanks for the feedback.

I've cleaned up the commit history and updated the messages as requested. I also added the test case for cluster-scoped resources.

Force-pushed the updates. Let me know if anything else is needed!

File: frontend/src/lib/k8s/api/v2/hooks.test.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (illume) [reply frontend/src/lib/k8s/api/v2/hooks.test.tsx:276]: Thanks! Marking as resolved.

**Comment by illume** [comment]
A little nit, we use Capitals at the start of the git subject after the context...
For example:
> frontend: multiplexer: Cancel pending timers in test teardown

**Comment by Kaniikura** [comment]
Updated commit subjects to follow the capitalization guideline after the prefix. No code changes.

**Comment by Kaniikura** [comment]
Thanks for the suggestion to ask in Slack!
 I’ve posted a review request in the Kubernetes #headlamp

File: frontend/src/lib/k8s/api/v2/hooks.ts [hunk 7]
```diff
@@ -214,46 +216,117 @@ export function useKubeObject<K extends KubeObject>({
...
+ * @param namespace - Optional namespace scope
+ * @returns The first endpoint that responds with an OK status
  *
- * @throws Error
- * When no endpoints are working
+ * @throws {ApiError}
+ * When every endpoint fails
  */
-const getWorkingEndpoint = async (
+const getFirstWorkingListEndpoint = async (
```
Comment (sniok) [inline frontend/src/lib/k8s/api/v2/hooks.ts]: getFirstWorkingListEndpoint  and getFirstWorkingGetEndpoint are essentially the same functions. could you introduce optional `name` parameter and turn it into just one function

Comment (Kaniikura) [reply frontend/src/lib/k8s/api/v2/hooks.ts]: Good catch, thanks! Combined both functions into one optional-name variant

File: frontend/src/lib/k8s/api/v2/hooks.ts [hunk 8]
```diff
@@ -214,46 +216,68 @@ export function useKubeObject<K extends KubeObject>({
...
+  const endpointsKey = useMemo(
+    () => endpoints.map(ep => `${ep.group ?? ''}/${ep.version}/${ep.resource}`),
+    [endpoints]
+  );
+
   const { data: endpoint, error } = useQuery<KubeObjectEndpoint, ApiError>({
     enabled: endpoints.length > 1,
-    queryKey: ['endpoints', endpoints],
-    queryFn: () => getWorkingEndpoint(endpoints, cluster!, namespace),
+    queryKey: ['endpoints', cluster, namespace, name ?? '', endpointsKey],
```
[the remark this answers is not in the document]
Comment (Kaniikura) [reply frontend/src/lib/k8s/api/v2/hooks.ts:277]: Yeah, this does create more cache entries. But for this flow correctness wins: endpoint probing is permission-sensitive (especially with RBAC/resourceNames), so sharing across names can cache the wrong result.

[the remark this answers is not in the document]
Comment (illume) [reply frontend/src/lib/k8s/api/v2/hooks.ts:277]: Thanks for the explanation.

**Comment by Kaniikura** [comment]
@sniok Just a quick nudge on this PR. I’ve incorporated the feedback. Could you take another look when you have a moment?
